lib/reporting/protocols_report.rb

```diff
@@ -0,0 +1,183 @@
+# frozen_string_literal: true
+
+require 'csv'
+begin
+  require 'reporting/cloudwatch_client'
+  require 'reporting/cloudwatch_query_quoting'
+  require 'reporting/command_line_options'
+rescue LoadError => e
+  warn 'could not load paths, try running with "bundle exec rails runner"'
+  raise e
+end
+
+module Reporting
+  class ProtocolsReport
+    include Reporting::CloudwatchQueryQuoting
+
+    attr_reader :time_range
+
+    SAML_AUTH_EVENT = 'SAML Auth'
+    OIDC_AUTH_EVENT = 'OpenID Connect: authorization request'
+
+    # @param [Range<Time>] time_range
+    def initialize(
+      issuers:,
+      time_range:,
+      verbose: false,
+      progress: false,
+      slice: 1.day,
+      threads: 10
+    )
+      @issuers = issuers
+      @time_range = time_range
+      @verbose = verbose
+      @progress = progress
+      @slice = slice
+      @threads = threads
+    end
+
+    def verbose?
+      @verbose
+    end
+
+    def progress?
+      @progress
+    end
+
+    def as_tables
+      [
+        overview_table,
+        protocols_table,
+        saml_signature_issues_table,
+      ]
+    end
+
+    def as_emailable_reports
+      [
+        Reporting::EmailableReport.new(
+          title: 'Overview',
+          table: overview_table,
+        ),
+        Reporting::EmailableReport.new(
+          title: 'State of Authentication',
+          table: protocols_table,
+        ),
+        Reporting::EmailableReport.new(
+          title: 'SAML Signature Issues',
+          table: saml_signature_issues_table,
+        ),
+      ]
+    end
+
+    def to_csvs
+      as_tables.map do |table|
+        CSV.generate do |csv|
+          table.each do |row|
+            csv << row
+          end
+        end
+      end
+    end
+
+    # @return Array<Hash>
+    def data
+      @data ||= begin
+        fetch_results
+      end
+    end
+
+    def fetch_results
+      cloudwatch_client.fetch(query:, from: time_range.begin, to: time_range.end)
+    end
+
+    def query
+      params = {
+        event: quote([SAML_AUTH_EVENT, OIDC_AUTH_EVENT]),
+      }
+
+      format(<<~QUERY, params)
+      fields name as protocol, properties.event_properties.service_provider as issuer, properties.event_properties.request_signed = 1 AS signed, properties.event_properties.request_signed != 1 AS not_signed, isempty(properties.event_properties.matching_cert_serial) and signed as invalid_signature
+      | filter name IN %{event}
+      | stats count(*) as request_count, sum(signed) as signed_count, sum(not_signed) as unsigned_count, sum(invalid_signature) as invalid_signature_count by protocol, issuer
+      QUERY
```

Comment:
I would consider formatting this for shorter lines for easier readability, also using consistent casing for keywords like `AS`
```suggestion
      format(<<~QUERY, params)
        fields
          name AS protocol
        , properties.event_properties.service_provider AS issuer
        , properties.event_properties.request_signed = 1 AS signed
        , properties.event_properties.request_signed != 1 AS not_signed
        , isempty(properties.event_properties.matching_cert_serial) AND signed AS invalid_signature
        | filter name IN %{event}
        | stats
          count(*) AS request_count
        , sum(signed) AS signed_count
        , sum(not_signed) AS unsigned_count
        , sum(invalid_signature) AS invalid_signature_count
          by
          protocol
        , issuer
      QUERY
```

Reply:
Personally, I don't like the leading commas - it hurts my eyes :-). However, if it is a standard here, I am happy to adopt it.

Reply:
Not 100% consistent but we do have many files formatted that way:
```bash
identity-idp> git grep -l  "  , " | wc -l
      18
```